syntactic-whitelist/src/test/scala/TestRunner.scala

```diff
@@ -0,0 +1,245 @@
+            val line = violation.tree.pos.startLine
+            acc.updated(line, acc.getOrElse(line, default = 0) + 1)
+          })
+
+        val allLinesZeroBased = (expectedViolationsCntZeroBasedLines.keys ++ violationsWithLines.keys).toList.sorted
+        val stringJoiner = new StringJoiner("\n")
+        for (lineZeroBased <- allLinesZeroBased) {
+          val expected = expectedViolationsCntZeroBasedLines.getOrElse(lineZeroBased, default = 0)
+          val actual = violationsWithLines.getOrElse(lineZeroBased, default = 0)
+          if (actual != expected){
+            stringJoiner.add(s"error at line ${lineZeroBased+1}: unexpected number of violations found: $actual, expected: $expected\n" +
+              s"\tfound violations are: ${invalid.violations.filter(_.tree.pos.startLine == lineZeroBased)}")
+          }
+        }
+        if (stringJoiner.length() > 0){
+          fail(stringJoiner.toString)
+        }
+      }
+    }
+
+    /**
+     * Fails the test if either the result of the check is not an instance of Invalid or if the reported violations
+     * do not conform to the pairs given as an argument
+     * @param expectedViolations pairs that associate each line to the expected number of violations on that line
+     *                              (lines with 0 expected violations can be omitted) <p>
+     *                              e.g. 5 -> 2, 7 -> 1 means 2 violations at line 5 and 1 at line 7
+     */
+    def expectInvalid(expectedViolations: (Int, Int)*): Unit = expectInvalid(expectedViolations.toMap)
+
+    /**
+     * Fails the test if either the result of the check is not an instance of Invalid or if the reported violations
+     * do not conform to the list of lines given as an argument
+     * @param expectedViolationLines each occurrence of a line index in this list is considered as an expected
+     *                               violation on that line <p>
+     *                               e.g. [5, 5, 7] means 2 violations at line 5 and 1 at line 7
+     */
+    def expectInvalidAtLines(expectedViolationLines: Int*): Unit = {
+      val expectedViolationsAsMap =
+        (for (line <- expectedViolationLines) yield line -> expectedViolationLines.count(_ == line)).toMap
+      expectInvalid(expectedViolationsAsMap)
+    }
+
+    /**
+     * @return an instance of TestRunner with the provided test execution steps
+     */
+    private def build(): TestRunner = {
+      if (filename.isEmpty) throw new IllegalStateException("cannot build TestRunner: no filename")
+      if (matcher.isEmpty) throw new IllegalStateException("cannot build TestRunner: no expectedRes")
+      if (features.isEmpty) throw new IllegalStateException("cannot build TestRunner: no features")
+      new TestRunner(
+        matcher.get,
+        filename.get,
+        features.toList,
+        dialect.getOrElse(DEFAULT_DIALECT),
+        testController,
+        uid
+      )
+    }
```

Comment:
I initially thought that this implementation offers more flexibility and makes the tests more readable. This was especially the case because previous versions of the tests had long lists of features to allow, but I agree that it makes less sense now. So you think that I should refactor the tests to simply use a function (e.g. `performTest`) that would take all the needed informations as arguments?

Reply:
If that’s possible then yes, I think that would be a win!

Reply:
Done in 62a07fcd901eeee07b0ef35c6e631aa1f3c13fc5